### ISSUE TITLE
[python-package] type hints in python package

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#3756]
## Request for Comment

#3581 officially ended support for Python 2 in the Python package.

Now that Python 2 support has been ended, I'd like to propose adding type hints to the library (https://docs.python.org/3/library/typing.html).

Benefits of adding type hints:

* serves as additional documentation about the types of arguments and return types of functions
* allows us to enable `mypy`, a static type checker for Python that will catch issues like "this function call can return None but you're accessing a property of the result as if it isn't None"
* allows other libraries that integrate with LightGBM (like `optuna`) to catch issues in their own code using `mypy`

I'm opening this Request for Comment to get thoughts on this narrow proposal:

> would you support a PR where I add type hints to the code in the https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/dask.py?

## Status: Accepted

~If maintainers agree to this proposal, we could add a "good first issue" about adding hints for other modules and could talk about `mypy` separately, at some point in the future.~

This was originally opened as a request for comment, but it's been accepted and we're now accepting contributions to add type hints in the Python package!

## How to contribute

Anyone is welcome to submit a pull request adding type hints to places in the Python package that are missing them!

Please limit pull requests to 1 function/method per pull request, to make them easier to review.

* ~`.ci/get_workflow_status.py`~ (shouldn't be touched because this file is executed by default Python 2 installed on GitHub Actions workers)
* [x] `docs/conf.py` (#4526)
* [x] `helpers/check_dynamic_dependencies.py` (#4382)
* [x] `helpers/parameter_generator.py` (#4474)
* [ ] `python-package/lightgbm/basic.py` (#7140)
* [x] `python-package/lightgbm/callback.py` (#4093)
* [x] `python-package/lightgbm/compat.py`
* [x] `python-package/lightgbm/dask.py` (#3866)
* [x] `python-package/lightgbm/engine.py`
* [x] `python-package/lightgbm/libpath.py` (#4070)
* [x] `python-package/lightgbm/plotting.py` (#4367)
* [ ] `python-package/lightgbm/sklearn.py` (#7140)
* [x] `python-package/setup.py` (#4376)

### Handling line breaks

If adding hints to a method or function greatly increases the length of its lines, break each argument onto a separate line and have the closing parenthesis begin its own line, vertically aligned with the keyword `def`. Like this:

```python
# before
def make_ranking(n_samples=100, n_features=20, n_informative=5, gmax=2,
                 group=None, random_gs=False, avg_gs=10, random_state=0):

# after
def make_ranking(
    n_samples: int = 100,
    n_features: int = 20,
    n_informative: int = 5,
    gmax: int = 2,
    group: Optional[List[int]] = None,
    random_gs: bool = False,
    avg_gs: int = 10,
    random_state: int = 0
) -> Tuple[pd.DataFrame, pd.DataFrame]:
```

### references that may help you

* pull request that added type hints in `python-package/lightgbm/dask.py`: https://github.com/microsoft/LightGBM/pull/3866
* Python official docs on typing: https://docs.python.org/3/library/typing.html
* related issue looking for help with fixing `mypy` warnings: #3868 

## How this improves LightGBM

Type hints in the code serve as additional documentation of the expected input and output types for functions and methods.

Having richer type hints can help static analyzers like `mypy` catch bugs that otherwise might not be caught by LightGBM's unit tests.

**Comment by StrikerRUS** [comment]
I'm +1.
This shouldn't be hard to add type hints, because for all public API functions (and even some internal ones) all types are already specified in docstrings.

**Comment by trivialfis** [comment]
xgboost has been adding type hints gradually and the dask module is completely typed.  From my experience it's quite useful in both documentation and type checks.

**Comment by jameslamb** [comment]
I did the following this morning:

* replaced "[RFC]" in the title with "[python-package]"
* removed label "question"
* added more details in the description to help contributors looking to submit PRs that add type hints

Anyone reading this is welcome to contribute! Thanks for helping to improve LightGBM.

**Comment by StrikerRUS** [comment]
I suggest moving to a scheme "1 PR - 1 function/method" for the rest files. There are 142 functions in `basic.py`. It's quite hard to review such huge PRs. Refer to #4367 for example.

**Comment by sagnik1511** [comment]
Hello @jameslamb, I want to contribute to the file **helpers/parameter_generator.py**. Can I update that??

**Comment by jameslamb** [comment]
@sagnik1511 yes please! You do not need to ask permission and are welcome to work on any file listed in the description of this issue.

**Comment by jameslamb** [comment]
This issue has been open for about 8 months now, so I think it's ok for maintainers to start picking up the work to finish it and not wait any longer for outside contributors to find it from searches for `good first issue`.

Anyone reading this is welcome to contribute if you'd like, as long as the issue is open, but I think maintainers should also start contributing as time permits, to improve the ability of `mypy` to catch issues (#3868).

**Comment by jameslamb** [comment]
I think `compat.py` can be safely ignored for this purpose, since it just includes  some try-catched imports for compatibility reasons.

**Comment by StrikerRUS** [comment]
I believe that during updating `python-package/lightgbm/sklearn.py` we need to replace `array-like` everywhere in docstrings with something meaningful. Originally arose in https://github.com/microsoft/LightGBM/pull/4547#discussion_r695997902. We need to use concrete data types to describe data sources for `X`, `y` and others. Not just because concrete types will be easier to understand for newcomers. But also because, according to the [official definition of `array-like` term](https://scikit-learn.org/stable/glossary.html#term-array-like) in scikit-learn, we use it incorrectly: we lose `scipy.sparse`, for example. Note, how `array-like` is used in `RandomForestRegressor`, for instance: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html#sklearn.ensemble.RandomForestRegressor.fit.

I'm posting this here instead of creating a new issue because I think these two tasks are very tight to each other. There is no `typing.ArrayLike` type, so one should explore all possible combinations of allowed types. After that there will be no problem to update a corresponding docstring in the same PR.

**Comment by makquel** [comment]
I'd like to contribute to this issue, are there any files remaining to static typing to be added?

**Comment by jameslamb** [comment]
Thanks @makquel ! There is plenty of work left in `basic.py` and `sklearn.py`. But please do not try to add hints for everything in either of those files. If you look through the things linked here, you'll see that we've been approaching this with many, small pull requests.

To start, could you try adding type hints to these functions? https://github.com/microsoft/LightGBM/blob/a5b2e049011da6a9b772116c5dde8174986b0bab/python-package/lightgbm/basic.py#L240-L277

**Comment by kunjshukla** [comment]
I would like to take this issue up if it's not closed

**Comment by jameslamb** [comment]
@kunjshukla sure, pull requests are welcome. Please make them small and focused... don't try to submit one large PR that adds type hints in all the remaining places missing them.
